contracts/EntryPoint.sol

```diff
@@ -401,5 +482,33 @@ contract EntryPoint is StakeManager {
         senderStorageCells = new uint256[](1);
         senderStorageCells[0] = cell;
     }
+
+    //save current free memory pointer
+    function getMemoryPointer() private pure returns (uint ptr) {
+        assembly {
+            ptr := mload(0x40)
+        }
+    }
+
+    //restore previously saved memory pointer.
+    // all "memory" allocations since getMemoryPointer are freed, and their memory will be re-uased.
+    function restoreMemoryPointer(uint ptr) private pure {
+        assembly {
+            mstore(0x40, ptr)
+        }
+    }
+
+    function copyOpsRange(
```

Comment:
I can see that you check that passed 'aggregator' is indeed the one trusted by the wallet in the `copyOpsRange` which is a separate loop of external calls, which is kind of wasteful given there is a `validateUserOp` call down the line. 

Even the loop itself is very weird - can't we save gas by switch to a two-dimensional array of UserOps or something like that?

Also, this is a very strange place to have this call anyways - copying is not related to this check.

Reply:
1. at first I wanted to "slice" the userOp array, but we can't, since it contains dynamic data, and we must copy the items one by one. Also note that we juggle 3 different layouts: the handleOp's incoming "calldata", the memory structure, and the outgoing "validateSignatures" memory (which is structured as the calldata for the called function). Even if these structures are exactly the same (in solidity), their layout differ, and requires one-by-one copy.
2. You're right, we can have a better API by an array of {userOp[], aggregator, signature}. It does add 2 zero-params to a "no aggregator" mode, but I Think that's minimal
3. the loop for call the wallets is required - the wallets are expected the EntryPoint to validate their entrypoint. otherwise, I can take some wallet and pass it with a "do-nothing aggregator" to the EntryPoint, and it will be "validated"
the other alternative is to add `validateUserOp` "aggregator" param, and the wallet will have to validate this field. This is not a replacement for the `getAggregator()` method, since at least off-chain, the bundler must be able to ask the wallet what aggregator it uses. 
Yes, tt will save the extra external call loop, so it seems good idea

Reply:
Re 1: Is there a way to arrange this data in a way that would avoid this shuffling of bytes? I assume this will save a lot of gas, especially for bigger bundles as they will pay much more for memory allocation. I may be wrong though.